op-node/rollup/derive/batch_queue.go

```diff
@@ -122,196 +97,140 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
-	// If not, just return a single batch
-	// Note: can't process a full epoch until we are closed
-	if bq.progress.Origin.Number >= epoch.Number+bq.config.SeqWindowSize && bq.progress.Closed {
-		bq.log.Info("Advancing full epoch", "origin", epoch, "tip", bq.progress.Origin)
-		// 2a. Gather all batches. First sort by timestamp and then by first seen.
-		var bns []uint64
-		for n := range bq.batchesByTimestamp {
-			bns = append(bns, n)
-		}
-		sort.Slice(bns, func(i, j int) bool { return bns[i] < bns[j] })
-
-		var batches []*BatchData
-		for _, n := range bns {
-			for _, batch := range bq.batchesByTimestamp[n] {
-				// Filter out batches that were submitted too late.
-				if uint64(batch.Batch.EpochNum)+bq.config.SeqWindowSize < batch.L1InclusionBlock.Number {
-					continue
-				}
-				// Pre filter batches in the correct epoch
-				if batch.Batch.EpochNum == rollup.Epoch(epoch.Number) {
-					batches = append(batches, batch.Batch)
-				}
-			}
+	l2SafeHead := bq.next.SafeL2Head()
+
+	if l2SafeHead.L1Origin != epoch.ID() {
+		return nil, NewResetError(fmt.Errorf("buffered L1 chain epoch %s in batch queue does not match safe head %s", epoch, l2SafeHead))
+	}
+
+	// Find the first-seen batch that matches all validity conditions.
+	// We may not have sufficient information to proceed filtering, and then we stop.
+	// There may be none: in that case we force-create an empty batch
+	nextTimestamp := l2SafeHead.Time + bq.config.BlockTime
+	var nextBatch *BatchWithL1InclusionBlock
+
+	// Go over all batches, in order of inclusion, and find the first batch we can accept.
+	// We filter in-place by only remembering the batches that may be processed in the future, or those we are undecided on.
+	var remaining []*BatchWithL1InclusionBlock
+	candidates := bq.batches[nextTimestamp]
+batchLoop:
+	for i, batch := range candidates {
+		validity := CheckBatch(bq.config, bq.log.New("batch_index", i), bq.l1Blocks, l2SafeHead, batch)
+		switch validity {
+		case BatchFuture:
+			return nil, NewCriticalError(fmt.Errorf("found batch with timestamp %d marked as future batch, but expected timestamp %d", batch.Batch.Timestamp, nextTimestamp))
+		case BatchDrop:
+			bq.log.Warn("dropping batch",
+				"batch_timestamp", batch.Batch.Timestamp,
+				"parent_hash", batch.Batch.ParentHash,
+				"batch_epoch", batch.Batch.Epoch(),
+				"txs", len(batch.Batch.Transactions),
+				"l2_safe_head", l2SafeHead.ID(),
+				"l2_safe_head_time", l2SafeHead.Time,
+			)
+			continue
+		case BatchAccept:
+			nextBatch = batch
+			// don't keep the current batch in the remaining items since we are processing it now,
+			// but retain every batch we didn't get to yet.
+			remaining = append(remaining, candidates[i+1:]...)
+			break batchLoop
+		case BatchUndecided:
+			remaining = append(remaining, batch)
+			bq.batches[nextTimestamp] = remaining
+			return nil, io.EOF
+		default:
+			panic(fmt.Errorf("unknown batch validity type: %d", validity))
```

Comment:
Exhaustive pattern matching is a nice feature to have in a language, the compiler would force this to be unreachable

Reply:
Rust vs Go :smile: